fem/integrator.hpp

```diff
@@ -0,0 +1,94 @@
+#ifndef MFEM_INTEGRATOR
+#define MFEM_INTEGRATOR
+
+#include "../config/config.hpp"
+#include "fe.hpp"
+
+namespace mfem
+{
+
+/** @brief This wraps a pointer to an integration rule, while allowing it to be designated as temporary, to be deleted after passing out of scope. */
+class IntegrationRulePtr
```

Comment:
Is this class really needed? I think we do not want ultimately a wrapper for a pointer of every type. That pointers from the standard library would be a better option and there can be one directly in `Integrator`, so something like `unique_ptr<const IntegrationRule> DefIntRule;` 😉 .

Reply:
It's not quite the right use-case for `unique_ptr`.  The returned pointer is usually to an integration rule owned by `IntRules`, which we do not want to go out-of-scope, but there's an exception where it may be a pointer to a temporary integration rule generated on-the-fly, which needs to be deleted by the calling scope.  So, it's really an occasionally-shared pointer, but going through and converting `IntRules` to store `shared_ptr`s to integration rules quickly grows in scope and also has some performance implications from the reference counting overhead (that is not necessary most of the time).  It is probably a design issue that this exception case requiring deletion by the caller exists, but I'm not sure what the best solution is.  The wrapper class here is slightly nicer than the previous pattern of taking a `bool&` that is updated to indicate whether the calling scope needs to `delete` the returned integration rule.  Would maybe a `std::variant<mfem::IntegrationRule*, std::unique_ptr<mfem::IntegrationRule>>` be a better way to get the desired conditional deletion without defining a new class?

Reply:
(Googling quickly, it looks like `std::variant` is only available since C++17, so I don't think it's permitted in MFEM yet.)

Reply:
It sounds like having a global variable `IntRules` is preventing usage of `std` smart pointers. Maybe we should change this in a future PR, since it would be a major change beyond the scope of this PR.

Reply:
Well, this starts to be very similar discussion to what we had with @pazner in #4091 . I think the best way forward would be to generalize by template `OperatorHandle` (to just `Handle` perhaps? 🙂 ). That would provide exactly the needed capacity for optionally owned pointers. How far we will adopt it in MFEM and/or extend it for reusing identical objects we may leave for later. Do you want to do it or should I? 🚀

Reply:
Yes, with regards to cleaning up ownership rules I think we should definitely start doing some modernization; my hope would be to this in a consistent way across the (very large) MFEM codebase, so it might take some discussion to land on an acceptable design.

I thought about this briefly, and also considered creating a `Handle<T>` class in MFEM that would _either_ contain a raw pointer (not owned) _or_ a `shared_ptr` (owned, of course).

Interfaces could be transitioned to taking `Handle<T>` instead of `T*` in a backwards-compatible way by having a non-owning constructor `Handle(T*)`.

This idea is sketched out here: https://godbolt.org/z/nnYxqe1ar

Reply:
That looks cool 😎 , but yes, that will require a bigger discussion, so for this PR, how about just a basic template replacing `unique_ptr` in the way the current implementation of `OperatorHandle` does? This could be extended to that `shared_ptr` later...

Reply:
It sounds like you're proposing a template generalization of `IntegrationRulePtr` that would work for other optionally-owned pointer types, and already has another possible use-case for operators.  I like that idea, although I'm less familiar with the other use case you're envisioning and not sure whether there are general conventions for where to put or how to implement these sorts of general "utility" types in MFEM, so maybe you can go ahead and implement the idea you already have if it's relatively easy.

Reply:
My vote on this (and #4091) is to try to reach a consensus about a systematic way to handle these and other ownership issues consistently in the code, rather than introduce one-off solutions that will have to be refactored later (maintaining backwards compatibility, etc.).

Reply:
Ok folks, I created #4569 with a basic templated generalization of `OperatorHandle`. It can be used here instead of the one-off class `IntegrationRulePtr`. I think it is a good solution for now and we may extend later following to what @pazner posted 😉 .

Reply:
Thanks for drafting an implementation of the new generalization.  One potential issue I see in the context of the present PR is that the proposed `Handle` interface doesn't look like it will transfer ownership when returning a `Handle` from a function, which is the behavior needed from `IntegrationRulePtr` here.  The concern is that if a `Handle` is created in a function (e.g., the `GetIntegrationRule` method here), then returned from that function, the returned value would be a shallow copy, and the original owning object would delete the pointer when destructed at the end of the function's scope.

Reply:
It is scoped in the `Integrator` anyway, so the idea (similar to [what I wrote above](https://github.com/mfem/mfem/pull/4557#discussion_r1830025862)) was to just make something like `Handle<const IntegrationRule> DefIntRule;` attribute of `Integrator` for now 😉 .

Reply:
@david-kamensky, is the approach I described at the beginning of https://github.com/mfem/mfem/pull/4569#issuecomment-2460971338 possible?

The idea would be that `NURBSMeshRules` would have an `IntegrationRules` data member. `NURBSMeshRules::GetElementRule` would return a reference to this data member; it would always be a non-owning reference (there would be no need for `deleteRule` — `deleteRule` would always be false). `NURBSMeshRules::GetElementRule` would resize/overwrite this internal integration rule if needed.

This way, all the `GetIntegrationRule` functions would return non-owning references/pointers, and there is no need for a "maybe-owning" pointer type.

Does that sound reasonable?

Reply:
Oh, okay, I missed that part of your suggestion.  My understanding of what you mean is that we'd just keep updating the value of this attribute with the `Handle::Reset` method when building temporary rules not owned by `IntRules`, which would resolve the issue I mentioned above.  In the case where a temporary rule is generated on each element, this would defer freeing the last temporary rule generated until the `Integrator` is destroyed (instead of until the assembly method completes), but that is probably not enough memory to be a significant concern.

Reply:
You are right, but it is the case only for the legacy OpenMP backend, which, as the name says, is not much used anymore and most of the integrators are not thread safe anyway (but I have not checked the specific integrators you modified). Though, as you are developing a general system, it would be better to support that, so there is preprocessor declaration `MFEM_THREAD_SAFE`, which is used to guard the shared attributes of the integrators (see e.g. `hyperbolic.cpp/hpp`). Maybe a solution is to pass the owning handle when `MFEM_THREAD_SAFE` is defined and non-owning one (pointing to the shared attribute) when it is not to keep some caching 😉 .

Reply:
Hi @david-kamensky, does https://github.com/mfem/mfem/pull/4557#discussion_r1831825377 sound possible to you? (If it's not clear what was meant there, I can provide a patch with the changes needed for the proposed approach). This unfortunately has the same thread-safety issue as what Jan suggested. The advantage is that no specialized types (neither `IntegrationRulePtr` nor `Handle`) would be needed.

Reply:
@pazner Yes, that sounds possible.  However, if we're okay with
* Lack of thread safety
* Deferring deletion of the last-used temporary element rule until after the scope of the element-level assembly method (until either the `Integrator` or `NURBSMeshRules` is deleted)

then I think we could also just go with @najlkin's original proposal of just using a `unique_ptr<IntegrationRule>` member of `Integrator`, which would avoid modifying the public interface of `NURBSMeshRules`.  (We actually couldn't combine the temporary-rule pointer with the existing non-owned `IntegrationRule* ir` member of `Integrator` as a single `Handle` even if we wanted to; whether to override that with a temporary rule depends on the element argument to the assembly method, and is not an inherent attribute of the `Integrator`.)

`NURBSMeshRules` is probably a more logical owner for the temporary rule, but I don't know how strict we want to be about preserving the existing interface with the `deleteRule` argument (to avoid "breaking changes"), even if it's not ideal.

Reply:
If the concern is changing the interface of `NURBSMeshRules`, the interface can remain the same, but `deleteRules` is just always set to false.

Reply:
If the only breaking change is to `NURBSMeshRules::GetElementRule`, I would think that is ok, since it is a relatively new function not used by anyone other than you @david-kamensky, as far as I know.

Reply:
But that thread safety is a trouble if we want to apply this system of integration rules everywhere... 🤔

Reply:
I like the idea of making a small breaking change to fix this issue of integration rule ownership at its source, by eliminating the `deleteRule` behavior from `NURBSMeshRules::GetElementRule`.  

We could have a (very memory-intensive) thread-safe fallback option of just pre-computing all the element rules and adding them to `NURBSMeshRules::elementToRule` in the `NURBSMeshRules` constructor, or simply put an `MFEM_ABORT` inside `#ifdef MFEM_THREAD_SAFE`/`#endif` in the code-path for using a temporary rule, which would only be reached if using `NURBSMeshRules` and multi-threading together.

Reply:
I pushed a new commit just now following the second approach of just `MFEM_ABORT`-ing if trying to create a temporary rule on a NURBS element with `MFEM_THREAD_SAFE` defined.  Otherwise, `NURBSMeshRules` has a `private` member `std::unique_ptr<const IntegrationRule> temporaryElementRule` that is re-assigned each time a new temporary rule is formed on an element (automatically `delete`-ing the previous rule).  

This completely removes the `IntegrationRulePtr` class and does not require any analogous replacement (e.g., `Handle`).  I made the small breaking change of removing the now-obsolete `bool& deleteRule` argument from `NURBSMeshRules::GetElementRule`.  (But, as @pazner pointed out, we could always leave it in as a do-nothing parameter if we want to strictly enforce backward compatibility of interfaces.)

Reply:
I would say I more or less like this solution. It kills some pathways, but legacy OpenMP in combination with NURBS with non-reduced rules are a special case of a special case, so not very likely to appear, I guess 😄 . Also it was not much supported before anyway, if I am not wrong.
We could think of a thread-safe approach with objects (temporary rules) assigned to the thread ids, [as I proposed elsewhere](https://github.com/mfem/mfem/pull/4065#issuecomment-2377434398), but that would be an overkill to do just specifically here. However, we may think about it in the context something more general like `Handle`... 🤔